Monika After Story/game/zz_submods.rpy

```diff
@@ -1,102 +1,472 @@
+
+    HEADER_GLOB = "**/header.json"
+    SUBMODS_DIR = "Submods"
+
+    SANE_STR_PATTERN = re.compile(r'^[a-zA-Z_\u00a0-\ufffd][ 0-9a-zA-Z_\u00a0-\ufffd]*$')
+
+
+    class _SubmodSchema(pydantic.BaseModel):
+        """
+        Schema for validating submod json
+        """
+        # NOTE: JSON specific:
+        header_version: int
+        # NOTE: Submod specific:
+        author: str
+        name: str
+        version: str
+        directory: str# NOTE: this isn't part of the json, will be added dynamically during loading
+        modules: tuple[str, ...]
+        description: str = ""
+        dependencies: dict[str, tuple[str, str]] = {}# pydantic handles mut args
+        settings_pane: str = ""
+        version_updates: dict[str, str] = {}# pydantic handles mut args
+        coauthors: tuple[str, ...] = ()
+        repository: str = ""
+        priority: int = 0
+
+        @pydantic.validator("header_version")
+        def validate_header_version(cls, value):
+            if value <= 0:
+                raise ValueError(
+                    f"Submod header version {value} is invalid"
+                )
+            if value < HEADER_VERSION:
+                raise ValueError(
+                    f"Submod header version {value} is outdated (expected {HEADER_VERSION})"
+                )
+            if value > HEADER_VERSION:
+                raise ValueError(
+                    f"Submod header version {value} is unknown (expected {HEADER_VERSION})"
+                )
+
+            return value
+
+        @pydantic.validator("author")
+        def validate_author(cls, value):
+            if re.match(SANE_STR_PATTERN, value) is None:
+                raise ValueError(f"Submod author '{value}' is invalid")
+
+            return value
+
+        @pydantic.validator("name")
+        def validate_name(cls, value):
+            if re.match(SANE_STR_PATTERN, value) is None:
+                raise ValueError(f"Submod name '{value}' is invalid")
+
+            return value
+
+        @pydantic.validator("version")
+        def validate_version(cls, value):
+            if not _is_valid_version(value):
+                raise ValueError(f"Submod version number '{value}' is invalid")
+
+            return value
+
+        @pydantic.validator("modules")
+        def validate_modules(cls, value, values):
+            if not value:
+                raise ValueError("Submod must define modules.")
+
+            # IMPORTANT: Sort in alpha
+            value = tuple(sorted(value))
+
+            submod_dir = values.get("directory", None)
+            if (
+                submod_dir is not None
+                and not _mas_loader.do_modules_exist(*(f"{submod_dir}/{m}" for m in value))
+            ):
+                raise ValueError(
+                    "One or more submod modules are missing: {}".format(
+                        ", ".join(map(lambda s: f"'{s}'", value))
+                    )
+                )
+
+            return value
```

Comment:
Do we want to change how we pass these in for usability? I understand we wish not to allow `/*` patterns as that effectively restores prior behaviour of unspecified files, however I feel it can be cumbersome to need to retype the full paths i.e.:
```json
{
    ...,
    "modules": [
        "main",
        "subfolder/file1",
        "subfolder/file2",
        "subfolder/file3"
    ]
}
```

or did we want to swap to a more literal (in terms of how the author would see their filestruct) format i.e.
```json
{
    ...,
    "modules": [
        "main",
        "subfolder": {
            "file1",
            "file2",
            "file3"
        }
    ]
}
```

Reply:
Couldn't get this format to work, leaving the current variant for the time being.